File: src/test/java/com/jacuum/web/SessionEndpointTest.java
package com.jacuum.web;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.jacuum.web.dto.CreateSessionRequest;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@SpringBootTest
@AutoConfigureMockMvc
class SessionEndpointTest {

    @Autowired MockMvc mvc;
    @Autowired ObjectMapper mapper;

    @Test void createSessionReturns200WithSessionId() throws Exception {
        var body = new CreateSessionRequest(null, "TINY", "Random", "Alice", "🤖", 100);
        mvc.perform(post("/api/session")
                .contentType(MediaType.APPLICATION_JSON)
                .content(mapper.writeValueAsString(body)))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.sessionId").isNotEmpty())
            .andExpect(jsonPath("$.status").value("SETUP"))
            .andExpect(jsonPath("$.map.width").isNumber());
    }

    @Test void startSessionReturns200() throws Exception {
        // create first
        var body = new CreateSessionRequest(null, "TINY", "Random", "Alice", "🤖", 50);
        String resp = mvc.perform(post("/api/session")
                .contentType(MediaType.APPLICATION_JSON)
                .content(mapper.writeValueAsString(body)))
            .andReturn().getResponse().getContentAsString();
        String id = mapper.readTree(resp).get("sessionId").asText();

        mvc.perform(post("/api/session/" + id + "/start"))
            .andExpect(status().isOk());
    }

    @Test void createSessionWithMissingAlgoNameReturns400() throws Exception {
        var body = new CreateSessionRequest(null, "TINY", null, "Alice", "🤖", 100);
        mvc.perform(post("/api/session")
                .contentType(MediaType.APPLICATION_JSON)
                .content(mapper.writeValueAsString(body)))
            .andExpect(status().isBadRequest());
    }

    @Test void createSessionWithMissingUsernameReturns400() throws Exception {
        var body = new CreateSessionRequest(null, "TINY", "Random", null, "🤖", 100);
        mvc.perform(post("/api/session")
                .contentType(MediaType.APPLICATION_JSON)
                .content(mapper.writeValueAsString(body)))
            .andExpect(status().isBadRequest());
    }

    @Test void createSessionWithMissingAvatarReturns400() throws Exception {
        var body = new CreateSessionRequest(null, "TINY", "Random", "Alice", null, 100);
        mvc.perform(post("/api/session")
                .contentType(MediaType.APPLICATION_JSON)
                .content(mapper.writeValueAsString(body)))
            .andExpect(status().isBadRequest());
    }

    @Test void createSessionWithBlankAlgoNameReturns400() throws Exception {
        var body = new CreateSessionRequest(null, "TINY", "  ", "Alice", "🤖", 100);
        mvc.perform(post("/api/session")
                .contentType(MediaType.APPLICATION_JSON)
                .content(mapper.writeValueAsString(body)))
            .andExpect(status().isBadRequest());
    }

    @Test void createSessionWithBlankUsernameReturns400() throws Exception {
        var body = new CreateSessionRequest(null, "TINY", "Random", "  ", "🤖", 100);
        mvc.perform(post("/api/session")
                .contentType(MediaType.APPLICATION_JSON)
                .content(mapper.writeValueAsString(body)))
            .andExpect(status().isBadRequest());
    }

    @Test void createSessionWithBlankAvatarReturns400() throws Exception {
        var body = new CreateSessionRequest(null, "TINY", "Random", "Alice", "  ", 100);
        mvc.perform(post("/api/session")
                .contentType(MediaType.APPLICATION_JSON)
                .content(mapper.writeValueAsString(body)))
            .andExpect(status().isBadRequest());
    }
}

File: src/test/kotlin/com/jacuum/web/SessionEndpointTest.kt
package com.jacuum.web

import com.fasterxml.jackson.databind.ObjectMapper
import com.jacuum.web.dto.CreateSessionRequest
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status

@SpringBootTest
@AutoConfigureMockMvc
class SessionEndpointTest @Autowired constructor(
    private val mvc: MockMvc,
    private val mapper: ObjectMapper
) {

    private fun createSession(body: Any) = mvc.perform(
        post("/api/session")
            .contentType(MediaType.APPLICATION_JSON)
            .content(mapper.writeValueAsString(body))
    )

    @Test fun createSessionReturns200WithSessionId() {
        createSession(CreateSessionRequest(null, "TINY", "Random", "Alice", "\uD83E\uDD16", 100))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.sessionId").isNotEmpty())
            .andExpect(jsonPath("$.status").value("SETUP"))
            .andExpect(jsonPath("$.map.width").isNumber())
    }

    @Test fun startSessionReturns200() {
        val resp = createSession(CreateSessionRequest(null, "TINY", "Random", "Alice", "\uD83E\uDD16", 50))
            .andReturn().response.contentAsString
        val id = mapper.readTree(resp).get("sessionId").asText()
        mvc.perform(post("/api/session/$id/start"))
            .andExpect(status().isOk())
    }

    @Test fun createSessionWithMissingAlgoNameReturns400() {
        createSession(CreateSessionRequest(null, "TINY", null, "Alice", "\uD83E\uDD16", 100))
            .andExpect(status().isBadRequest())
    }

    @Test fun createSessionWithMissingUsernameReturns400() {
        createSession(CreateSessionRequest(null, "TINY", "Random", null, "\uD83E\uDD16", 100))
            .andExpect(status().isBadRequest())
    }

    @Test fun createSessionWithMissingAvatarReturns400() {
        createSession(CreateSessionRequest(null, "TINY", "Random", "Alice", null, 100))
            .andExpect(status().isBadRequest())
    }

    @Test fun createSessionWithBlankAlgoNameReturns400() {
        createSession(CreateSessionRequest(null, "TINY", "  ", "Alice", "\uD83E\uDD16", 100))
            .andExpect(status().isBadRequest())
    }

    @Test fun createSessionWithBlankUsernameReturns400() {
        createSession(CreateSessionRequest(null, "TINY", "Random", "  ", "\uD83E\uDD16", 100))
            .andExpect(status().isBadRequest())
    }

    @Test fun createSessionWithBlankAvatarReturns400() {
        createSession(CreateSessionRequest(null, "TINY", "Random", "Alice", "  ", 100))
            .andExpect(status().isBadRequest())
    }
}
